src/CSE.cpp

```diff
@@ -84,17 +73,17 @@ class GVN : public IRMutator {
 public:
     struct Entry {
         Expr expr;
-        int use_count;
+        int use_count = 0;
+        // All consumer Exprs for which this is the last child Expr.
+        map<ExprWithCompareCache, int> uses;
+        Entry(const Expr &e) : expr(e) {}
     };
     vector<Entry> entries;
 
-    typedef map<ExprWithCompareCache, int> CacheType;
-    CacheType numbering;
-
-    map<Expr, int, ExprCompare> shallow_numbering;
+    map<const BaseExprNode *, int> shallow_numbering, output_numbering;
```

Comment:
I will bet you a dollar that we need to use `IRHandle` here (instead of BaseExprNode*) to avoid possible refcounting bugs.